### PR TITLE
Add table_aws_cloudwatch_alarm. Closes #128

### LINKED ISSUE #128: Add table_aws_cloudwatch
**Describe the new table**
Need to query the CloudWatch to get the list of alarms and associated metrics in it. 

**References**

We need to achieve some of the steps as per CIS recommendation 
4.1 Ensure a log metric filter and alarm exist for unauthorised API calls(Automated) - The  steps 5 & 6 
4.2 Ensure a log metric filter and alarm exist for Management Console sign-in without MFA (Automated) - The  steps 5 & 6 
+ most of the section 4

### PR DESCRIPTION
**Pull request by khushboo9024**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```

SETUP: tests/aws_clouwatch_alarm []

PRETEST: tests/aws_clouwatch_alarm

TEST: tests/aws_clouwatch_alarm
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_partition.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_cloudwatch_metric_alarm.named_test_resource: Creating...
aws_cloudwatch_metric_alarm.named_test_resource: Creation complete after 4s [id=turbottest74468]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

account_id = 013122550996
aws_partition = aws
region_name = us-east-1
resource_aka = arn:aws:cloudwatch:us-east-1:013122550996:alarm:turbottest74468
resource_name = turbottest74468

Running SQL query: test-get-query.sql
[
  {
    "alarm_arn": "arn:aws:cloudwatch:us-east-1:013122550996:alarm:turbottest74468",
    "alarm_description": "turbottest74468",
    "comparison_operator": "GreaterThanOrEqualToThreshold",
    "metric_name": "CPUUtilization",
    "name": "turbottest74468"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "arn:aws:cloudwatch:us-east-1:013122550996:alarm:turbottest74468"
    ],
    "alarm_arn": "arn:aws:cloudwatch:us-east-1:013122550996:alarm:turbottest74468",
    "alarm_description": "turbottest74468",
    "comparison_operator": "GreaterThanOrEqualToThreshold",
    "metric_name": "CPUUtilization",
    "name": "turbottest74468",
    "tags": {
      "name": "turbottest74468"
    },
    "threshold": 80
  }
]
✔ PASSED

Running SQL query: test-list-call-query.sql
[
  {
    "alarm_arn": "arn:aws:cloudwatch:us-east-1:013122550996:alarm:turbottest74468",
    "alarm_description": "turbottest74468",
    "comparison_operator": "GreaterThanOrEqualToThreshold",
    "metric_name": "CPUUtilization",
    "name": "turbottest74468"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "account_id": "013122550996",
    "akas": [
      "arn:aws:cloudwatch:us-east-1:013122550996:alarm:turbottest74468"
    ],
    "alarm_arn": "arn:aws:cloudwatch:us-east-1:013122550996:alarm:turbottest74468",
    "name": "turbottest74468",
    "partition": "aws",
    "region": "us-east-1",
    "tags": {
      "name": "turbottest74468"
    },
    "title": "turbottest74468"
  }
]
✔ PASSED

POSTTEST: tests/aws_clouwatch_alarm

TEARDOWN: tests/aws_clouwatch_alarm

SUMMARY:

1/1 passed.


```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```

# Basic info

```sql
select
  name,
  state_value,
  metric_name,
  actions_enabled,
  comparison_operator,
  namespace,
  statistic
from
  aws_cloudwatch_alarm;

+------------------------------------------------------------------------------------------+-------------------+-------------------------------+-----------------+-------------------------------+----------------------+-----------+
| name                                                                                     | state_value       | metric_name                   | actions_enabled | comparison_operator           | namespace            | statistic |
+------------------------------------------------------------------------------------------+-------------------+-------------------------------+-----------------+-------------------------------+----------------------+-----------+
| turbottest25891                                                                          | INSUFFICIENT_DATA | CPUUtilization                | true            | GreaterThanOrEqualToThreshold | AWS/EC2              | Average   |
| TargetTracking-table/table1-AlarmHigh-a3b9103f-7bd0-4d88-9733-6ef3cd160ddb               | OK                | ConsumedReadCapacityUnits     | true            | GreaterThanThreshold          | AWS/DynamoDB         | Sum       |
| Default_Test_Alarm5                                                                      | INSUFFICIENT_DATA | <null>                        | true            | GreaterThanOrEqualToThreshold | <null>               | <null>    |
| TargetTracking-table/table1-ProvisionedCapacityLow-1cbc2d41-2aaf-43d6-affb-78c65b7f9ee3  | OK                | ProvisionedReadCapacityUnits  | true            | LessThanThreshold             | AWS/DynamoDB         | Average   |
| cis-alarm                                                                                | INSUFFICIENT_DATA | metrif-filter-test0005        | true            | GreaterThanOrEqualToThreshold | LogMetrics           | Sum       |
| cis0302-alarm                                                                            | INSUFFICIENT_DATA | cis0302                       | true            | GreaterThanThreshold          | LogMetrics           | Sum       |
| Billing alarm                                                                            | ALARM             | EstimatedCharges              | true            | GreaterThanThreshold          | AWS/Billing          | Maximum   |
| Default_Test_Alarm4                                                                      | INSUFFICIENT_DATA | <null>                        | true            | GreaterThanOrEqualToThreshold | <null>               | <null>    |
| Default_Test_Alarm3                                                                      | INSUFFICIENT_DATA | Default_Test                  | true            | GreaterThanOrEqualToThreshold | CW EXAMPLE METRICS   | Average   |
| alarm-cis0307                                                                            | INSUFFICIENT_DATA | cis0307                       | true            | GreaterThanThreshold          | LogMetrics           | Sum       |
| alarm-cis0313                                                                            | INSUFFICIENT_DATA | cis0313                       | true            | GreaterThanThreshold          | LogMetrics           | Sum       |
| metricfilter005-alarm                                                                    | INSUFFICIENT_DATA | metrif-filter-test0005        | true            | GreaterThanOrEqualToThreshold | LogMetrics           | Sum       |
| alarm-cis0314                                                                            | INSUFFICIENT_DATA | cis0314                       | true            | GreaterThanThreshold          | LogMetrics           | Sum       |
| TargetTracking-table/table1-AlarmHigh-b293885f-b438-494e-8817-87154da80a30               | OK                | ConsumedWriteCapacityUnits    | true            | GreaterThanThreshold          | AWS/DynamoDB         | Sum       |
| alarm-cis0310                                                                            | INSUFFICIENT_DATA | cis0310                       | true            | GreaterThanThreshold          | LogMetrics           | Sum       |
| testing1                                                                                 | INSUFFICIENT_DATA | MissingBlocks                 | true            | GreaterThanThreshold          | AWS/ElasticMapReduce | <null>    |


```
# List of cloudwatch alarms whose state is in alarm

```sql
select
  name,
  alarm_arn,
  state_value,
  state_reason
from
  aws_cloudwatch_alarm
where
 state_value = 'ALARM';


+---------------------------------------------------------------------------------------------------------------------------+-------------+---------------------------------------------------------------------------------------------------
| alarm_arn                                                                                                                 | state_value | state_reason                                                                                      
+---------------------------------------------------------------------------------------------------------------------------+-------------+---------------------------------------------------------------------------------------------------
| arn:aws:cloudwatch:us-east-1:013122550996:alarm:TargetTracking-table/table1-AlarmLow-4932b9b9-c9be-4a0f-a538-f6b128eb7134 | ALARM       | Threshold Crossed: 15 datapoints were less than the threshold (150.0). The most recent datapoints 
| arn:aws:cloudwatch:us-east-1:013122550996:alarm:TargetTracking-table/table1-AlarmLow-2464c435-817a-4d0f-ad8a-5fcd2def54d8 | ALARM       | Threshold Crossed: 15 datapoints were less than the threshold (150.0). The most recent datapoints

```

# List of cloudwatch alarms whose action enabled is on

```sql
select
  alarm_arn,
  actions_enabled
from
  aws_cloudwatch_alarm
where
 actions_enabled;



+------------------------------------------------------------------------------------------------------------------------------------------+-----------------+
| alarm_arn                                                                                                                                | actions_enabled |
+------------------------------------------------------------------------------------------------------------------------------------------+-----------------+
| arn:aws:cloudwatch:us-east-2:013122550996:alarm:turbottest25891                                                                          | true            |
| arn:aws:cloudwatch:us-east-1:013122550996:alarm:Billing alarm                                                                            | true            |
| arn:aws:cloudwatch:us-east-1:013122550996:alarm:Default_Test_Alarm4                                                                      | true            |
| arn:aws:cloudwatch:us-east-1:013122550996:alarm:Default_Test_Alarm5                                                                      | true            |
| arn:aws:cloudwatch:us-east-1:013122550996:alarm:Default_Test_Alarm3                                                                      | true            |
| arn:aws:cloudwatch:us-east-1:013122550996:alarm:alarm-cis0307                                                                            | true            |
| arn:aws:cloudwatch:us-east-1:013122550996:alarm:TargetTracking-table/table1-AlarmHigh-a3b9103f-7bd0-4d88-9733-6ef3cd160ddb               | true            |
| arn:aws:cloudwatch:us-east-1:013122550996:alarm:TargetTracking-table/table1-ProvisionedCapacityHigh-badd920a-f473-41d6-ba46-acd9fe51df6d | true            |
```

# Metric attached to cloudwatch alarm based on a single metric

```sql
select
  name,
  metric_name,
  namespace,
  period,
  statistic,
  dimensions
from
  aws_cloudwatch_alarm
where
  metric_name is not null;


+------------------------------------------------------------------------------------------+-------------------------------+----------------------+--------+-----------+---------------------------------------------------------------------+
| name                                                                                     | metric_name                   | namespace            | period | statistic | dimensions                                                          |
+------------------------------------------------------------------------------------------+-------------------------------+----------------------+--------+-----------+---------------------------------------------------------------------+
| turbottest25891                                                                          | CPUUtilization                | AWS/EC2              | 120    | Average   | <null>                                                              |
| alarm-cis0306                                                                            | cis0306                       | LogMetrics           | 300    | Sum       | <null>                                                              |
| alarm-cis0305                                                                            | cis0305                       | LogMetrics           | 300    | Sum       | <null>                                                              |
| Billing alarm                                                                            | EstimatedCharges              | AWS/Billing          | 21600  | Maximum   | [{"Name":"Currency","Value":"USD"}]                                 |
| TargetTracking-table/table1-ProvisionedCapacityHigh-7aaf155b-40e3-4961-aca3-ddd7e4c30ebb | ProvisionedReadCapacityUnits  | AWS/DynamoDB         | 300    | Average   | [{"Name":"TableName","Value":"table1"}]                             |
```

# List of metrics attached to cloudwatch alarm based on a metric math expression

```sql
select
   name,
  metric ->> 'Id' as metric_id,
  metric ->> 'Expression' as metric_expression,
  metric -> 'MetricStat' -> 'Metric' ->> 'MetricName' as metric_name,
  metric -> 'MetricStat' -> 'Metric' ->> 'Namespace' as metric_namespace,
  metric -> 'MetricStat' -> 'Metric' ->> 'Dimensions' as metric_dimensions,
  metric ->> 'ReturnData' as metric_return_data
from
  aws_cloudwatch_alarm,
  jsonb_array_elements(metrics) as metric;


+---------------------+-----------+-------------------+--------------------+------------------+-------------------+--------------------+
| name                | metric_id | metric_expression | metric_name        | metric_namespace | metric_dimensions | metric_return_data |
+---------------------+-----------+-------------------+--------------------+------------------+-------------------+--------------------+
| Default_Test_Alarm4 | m1        | <null>            | ConnectionsFailed  | MyService        | <null>            | true               |
| Default_Test_Alarm5 | m1        | <null>            | ConnectionsFailed  | MyService        | <null>            | true               |
| Default_Test_Alarm5 | m2        | <null>            | ConnectionsDropped | MyService        | <null>            | false              |
+---------------------+-----------+-------------------+--------------------+------------------+-------------------+--------------------+

### CODE REVIEW

File: aws/table_aws_cloudwatch_alarm.go [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+				Name:        "metric_name",
+				Description: "The name of the metric associated with the alarm, if this is an alarm based on a single metric.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "namespace",
+				Description: "The namespace of the metric associated with the alarm.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
```
Comment (Subhajit97) [inline aws/table_aws_cloudwatch_alarm.go:94]: Move all the JSON type columns at the end.

File: aws-test/tests/aws_cloudwatch_alarm/variables.tf [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  metric_name = "CPUUtilization"
+  namespace = "AWS/EC2"
+  period = 120
+  statistic = "Average"
+  threshold = 80
+  alarm_description = var.resource_name
+  tags = {
+    name = var.resource_name
+  }
+}
```
Comment (Subhajit97) [inline aws-test/tests/aws_cloudwatch_alarm/variables.tf:64]:
```suggestion
}

```

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
Resolve conflicts and make suggested changes

**Review by Subhajit97 — APPROVED** [review]
LGTM

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**